src/book/02-projects/01-nubots/03-build-system.mdx

```diff
@@ -5,3 +5,265 @@ title: Build System
+
+- `fvec3`: A vector with 3 components, maps to an `Eigen::Vector3f`
+- `vec3`: A vector with 3 components, maps to an `Eigen::Vector3d`
+- `mat3`: A 3x3 matrix, maps to an `Eigen::Matrix3d`
+
+All messages are packaged in a similar fashion to the C++ namespaces used for modules. For instance, `Image.proto` is in the `message.input` package.
+
+NUClear Roles creates a C++ wrapper class for each message to create a simpler interface for accessing and mutating message fields. The C++ namespace for a message follows its package. For instance, the `message.input.Image` message will have a C++ namespace of `message::input::Image`.
+
+#### Utilities
+
+Utilities are collections of functions that perform useful tasks. Typically, these functions are stateless and are (or feasibly could be) used by multiple different modules (or even other utilities).
+
+As is the case with modules and messages, utilities are grouped together into meta-utilities. For instance, utilities dealing with angles, coordinate system transforms, and geometrical constructs live in the `shared/utility/math` meta-utility.
+
+#### Tools and the `b` command
+
+NUClear Roles introduces the `b` command that wraps up all of the functionality needed to create modules and build the NUbots code.
+
+| Command   | Description                                       |
+| --------- | ------------------------------------------------- |
+| configure | Configure the CMake system for the current target |
+| build     | Build the currently configured code               |
+| edit      | Edit configuration files                          |
+| format    | Applies code formatting rules to codebase         |
+| install   | Install build code to specified robot             |
+| module    | Generate new modules                              |
+| run       | Run a role locally                                |
+| target    | Select a target to work on                        |
+
+##### Configure
+
+`./b configure` will ensure that a build folder is created in the current Docker image and will then run `cmake` on the source code ensuring that CMake creates all of the necessary build files inside of the build folder.
+
+It is possible to specify extra arguments to `cmake` to control the configuration process. For example, to enable the building of unit tests one would run
+
+```sh
+./b configure -- -DBUILD_TESTS=ON
+```
+
+---
+
+**NOTE**
+`--` is needed to prevent python from trying to process the argument itself
+
+---
+
+Passing `-i` to `./b configure` will open up an interactive window with `ccmake` so that you may modify the CMake configuration.
+
+##### Build
+
+As one might expect, `./b build` will build all enabled modules using `ninja`.
+
+If unit tests are enabled in the CMake configuration (`-DBUILD_TESTS=ON`) then `./b build test` can be run (after first running `/b build`) to execute all of the unit tests.
```

Comment:
Actually running ./b build test will build and then execute the tests anyway. I also might add that you can build individual roles by using e.g. `./b build sensortest`